### PR TITLE
Initial Drain Cleaner contribution

### PR DESCRIPTION
**Pull request by scholzj**
This PR contains the initial contribution of the Drain Cleaner utility as approved in [SP-029](https://github.com/strimzi/proposals/blob/main/029-adopt-the-drain-cleaner-utility.md). This contains only the preexisting code. In next PRs, we will need to work on the CI integration, push the container images into the Strimzi Quay.io org etc.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,168 @@
 [![License](https://img.shields.io/badge/license-Apache--2.0-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0)
 [![Twitter Follow](https://img.shields.io/twitter/follow/strimziio.svg?style=social&label=Follow&style=for-the-badge)](https://twitter.com/strimziio)
 
-# Strimzi Template repository
+# Strimzi Drain Cleaner
 
-This repository is currently empty
+Strimzi Drain Cleaner is utility which helps with moving the Kafka pods deployed by [Strimzi](https://strimzi.io/) from nodes which are being drained.
```
Comment (ppatierno) [inline README.md:6]:
```suggestion
The Strimzi Drain Cleaner is an utility which helps with moving the Kafka pods deployed by [Strimzi](https://strimzi.io/) from Kubernetes nodes which are being drained.
```

File: README.md [hunk 2]
```diff
@@ -1,6 +1,168 @@
...
+It is useful if you want the Strimzi operator to move the pods instead of Kubernetes itself.
+The advantage of this approach is that the Strimzi operator makes sure that no pods become under-replicated during the node draining.
+To use it:
+
+* Deploy Kafka using Strimzi and configure the `PodDisruptionBudgets` for Kafka and Zookeeper to have `maxUnavailable` set to `0`.
```
Comment (ppatierno) [inline README.md:11]:
```suggestion
* Deploy Kafka using Strimzi and configure the `PodDisruptionBudgets` for Kafka and ZooKeeper to have `maxUnavailable` set to `0`.
```

File: README.md [hunk 3]
```diff
@@ -1,6 +1,168 @@
...
+    template:
+      podDisruptionBudget:
+        maxUnavailable: 0
+  entityOperator:
+    topicOperator: {}
+    userOperator: {}
+```
+
+* Deploy the Strimzi Drain Cleaner
+* Drain the node with some Kafka or Zookeeper pods using the `kubectl drain` command
```
Comment (ppatierno) [inline README.md:60]:
```suggestion
* Drain the node with some Kafka or ZooKeeper pods using the `kubectl drain` command
```

Comment (ppatierno) [reply README.md:60]: Maybe also adding a link to the Kubernetes draining command, i.e. https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/

File: README.md [hunk 4]
```diff
@@ -1,6 +1,168 @@
...
+
+## How does it work?
+
+Strimzi Drain Cleaner using Kubernetes Admission Control features and Validating Web-hooks to find out when something tries to evict the Kafka or Zookeeper pods are.
```
Comment (ppatierno) [inline README.md:64]:
```suggestion
The Strimzi Drain Cleaner uses Kubernetes Admission Control features and Validating Web-hooks to find out when something tries to evict the Kafka or ZooKeeper pods.
```

File: README.md [hunk 5]
```diff
@@ -1,6 +1,168 @@
...
+
+### On Kubernetes without CertManager
+
+On Kubernetes, when you do not use Cert Manager, the certificates needed for the web-hook need to be generated manually.
+Follow the instructions in `./deploy/kubernetes` directory.
+
+## See it in action
+
+You can easily test how it works:
+* Install Strimzi or AMQ Streams in your cluster
```
Comment (ppatierno) [inline README.md:99]:
```suggestion
* Install Strimzi in your cluster
```

File: README.md [hunk 6]
```diff
@@ -1,6 +1,168 @@
...
+* Deploy Kafka cluster with Pod Disruption Budget configuration having `masUnavailable` set to `0` as shown in the example above
```
Comment (ppatierno) [inline README.md:100]:
```suggestion
* Deploy Kafka cluster with Pod Disruption Budget configuration having `maxUnavailable` set to `0` as shown in the example above
```

File: README.md [hunk 7]
```diff
@@ -1,6 +1,168 @@
...
+```shell script
+./mvnw package -Pnative
+```
+
+Or you can run the native executable build for Linux in a container using: 
+```shell script
+./mvnw package -Pnative -Dquarkus.native.container-build=true
+```
+
+You can then execute your native executable with: `./target/strimzi-cistic-odpadu-1.0.0-SNAPSHOT-runner`.
```
Comment (ppatierno) [inline README.md:134]: wondering if the artifact should be `strimzi-drain-cleaner`

File: deploy/certmanager/060-Deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      app: strimzi-drain-cleaner
+  template:
+    metadata:
+      labels:
+        app: strimzi-drain-cleaner
+    spec:
+      serviceAccountName: strimzi-drain-cleaner
+      containers:
+        - name: strimzi-drain-cleaner
+          image: quay.io/scholzj/strimzi-cistic-odpadu:latest
```
Comment (ppatierno) [inline deploy/certmanager/060-Deployment.yaml:21]: Just as a remainder to update the image field to the official image in the Strimzi repo that will come later.

File: deploy/certmanager/070-ValidatingWebhookConfiguration.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        operations:  ["CREATE"]
+        resources:   ["pods/eviction"]
+        scope:       "Namespaced"
+    clientConfig:
+      service:
+        namespace: "strimzi-drain-cleaner"
+        name: "strimzi-drain-cleaner"
+        path: /drainer
+        port: 443
+      caBundle: Cg==
```
Comment (ppatierno) [inline deploy/certmanager/070-ValidatingWebhookConfiguration.yaml:23]: What's this `Cg==` value here as CA bundle?

File: deploy/certmanager/060-Deployment.yaml (hunk 8 above)
Comment (scholzj) [reply deploy/certmanager/060-Deployment.yaml:21]: That needs to be done when we have the CI and are actually pushing the images into the Strimzi repositories.

File: deploy/certmanager/070-ValidatingWebhookConfiguration.yaml (hunk 9 above)
Comment (scholzj) [reply deploy/certmanager/070-ValidatingWebhookConfiguration.yaml:23]: This is an empty string in base64. Cert Manager updates it after the deployment.

File: README.md (hunk 7 above)
Comment (scholzj) [reply README.md:134]: This contributes the original code. We should change it in the future.

File: deploy/certmanager/070-ValidatingWebhookConfiguration.yaml (hunk 9 above)
Comment (ppatierno) [reply deploy/certmanager/070-ValidatingWebhookConfiguration.yaml:23]: Right!

**Review by ppatierno — APPROVED** [review]

**Pull request merged**